### PR TITLE
Service worker

### PR DESCRIPTION
**Pull request by pbywater**
- Adds service worker
- Caches most of the files (still an issue with the font)
- Saves data to render nodes offline
- User can remove a memory whilst offline
- Memory will delete from database when user goes back online

### CODE REVIEW

**Review by ghost — APPROVED** [review]

File: src/client/d3/d3.js [hunk 1]
```diff
@@ -1,38 +1,48 @@
-const { initTagMenu, showDeleteButton, hoveringOnDelete, hideDeleteButton, initSubmitMemory, tagSorting, constructTagList, showHeading } = require('../helpers/helpers.js');
+const { initTagMenu, showDeleteButton, hoveringOnDelete, hideDeleteButton, initSubmitMemory, tagSorting, constructTagList, showHeading, saveMemoryIdToStorage, removeMemoryFromStoredData, removeMemoriesDeletedOffline } = require('../helpers/helpers.js');
 const { animationDuration, width, height, jsonUrl, svg, fdGrp, nodeGrp, linkGrp } = require('./setup.js');
 const { sortWithMax, binByTag, centralMaxNodesByTag, memoryNodesAndLinks } = require('../node_transformations');
 const { appendPopUp, randomPopUp } = require('./modals.js');
 const { newUserIntro } = require('./newUserIntro.js');
 
+if ('serviceWorker' in navigator) {
```
Comment (des-des) [inline src/client/d3/d3.js]: Put this in your main js file!

File: public/service-worker.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
+/* eslint-disable */
```
Comment (des-des) [inline public/service-worker.js:1]: Can you link the me where you got this file

Comment (des-des) [reply public/service-worker.js:1]: I mean, drop a link in as a comment

File: src/client/d3/d3.js [hunk 3]
```diff
@@ -224,12 +234,19 @@ function render(updatedData) {
...
-        success: update,
-      });
+      if (navigator.onLine) {
+        $.ajax({
+          method: 'DELETE',
+          url: 'memories',
+          data: { id },
+          success: update,
+        });
+        render(formatData(data));
```
Comment (des-des) [inline src/client/d3/d3.js]: this is called by update, do you want to run it twice?

File: src/client/helpers/helpers.js [hunk 4]
```diff
@@ -220,6 +220,59 @@ function constructTagList(data) {
   </li>`);
 }
 
+function saveMemoryIdToStorage(id) {
+  if (localStorage.getItem('toDelete') !== null) {
+    const memoriesWaitingToBeRemoved = JSON.parse(localStorage.getItem("toDelete"));
+    memoriesWaitingToBeRemoved.memories.push(id);
+    const saveMemoriesToDelete = JSON.stringify(memoriesWaitingToBeRemoved);
+    localStorage.toDelete = saveMemoriesToDelete;
```
Comment (des-des) [inline src/client/helpers/helpers.js]: localStorage.setItem('toDelete' ?

File: src/client/helpers/helpers.js [hunk 5]
```diff
@@ -220,6 +220,59 @@ function constructTagList(data) {
...
+  }
+  else {
+    const saveDeletedMemory = JSON.stringify({memories: [id]});
```
Comment (des-des) [inline src/client/helpers/helpers.js:233]: Hmm, not sure this works, should be an array `[{memories: [id]}]`

File: src/client/helpers/helpers.js [hunk 6]
```diff
@@ -220,6 +220,59 @@ function constructTagList(data) {
...
+    localStorage.toDelete = saveDeletedMemory;
```
Comment (des-des) [inline src/client/helpers/helpers.js:234]: setItem

File: src/client/helpers/helpers.js [hunk 7]
```diff
@@ -220,6 +220,59 @@ function constructTagList(data) {
...
+
+function removeMemoryFromStoredData(id) {
+  const offlineData = JSON.parse(localStorage.getItem('data'));
+  offlineData.forEach((memory, index) => {
+    if (memory.id == id) {
+      offlineData.splice(index, 1);
+    }
+  });
+  const offlineDataAfterRemoving = JSON.stringify(offlineData);
+  localStorage.data = offlineDataAfterRemoving;
```
Comment (des-des) [inline src/client/helpers/helpers.js]: setItem

File: src/client/helpers/helpers.js [hunk 8]
```diff
@@ -220,6 +220,59 @@ function constructTagList(data) {
...
+return offlineData;
+}
+
+function removeMemoryFromLocalStorage(index) {
+  const memoriesWaitingToBeRemoved = JSON.parse(localStorage.getItem("toDelete"));
+  if (memoriesWaitingToBeRemoved.memories.length === 1) {
+    localStorage.removeItem('toDelete');
+  }
+  else {
+  memoriesWaitingToBeRemoved.memories.splice(index, 1);
```
Comment (des-des) [inline src/client/helpers/helpers.js]: indentation

File: src/client/helpers/helpers.js [hunk 9]
```diff
@@ -220,6 +220,59 @@ function constructTagList(data) {
   </li>`);
 }
 
+function saveMemoryIdToStorage(id) {
```
Comment (des-des) [inline src/client/helpers/helpers.js]: This name is confusing.

`setMemoryDeletePending`?

File: src/client/helpers/helpers.js (hunk 4 above)
Comment (des-des) [reply src/client/helpers/helpers.js]: We can refac this
```js
const toDeleteJSON = localStorage.getItem('toDelete')
const toDelete = (toDeleteJSON ? JSON.parse(toDeleteJSON) : []).concat(id)
localStorage.setItem(
  'toDelete',
  JSON.stringify(toDelete)
)
```

File: src/client/helpers/helpers.js [hunk 10]
```diff
@@ -220,6 +220,59 @@ function constructTagList(data) {
...
+    if (memory.id == id) {
+      offlineData.splice(index, 1);
+    }
+  });
+  const offlineDataAfterRemoving = JSON.stringify(offlineData);
+  localStorage.data = offlineDataAfterRemoving;
+return offlineData;
+}
+
+function removeMemoryFromLocalStorage(index) {
```
Comment (des-des) [inline src/client/helpers/helpers.js]: `cancelMemoryPendingDelete`?

File: src/client/helpers/helpers.js [hunk 11]
```diff
@@ -220,6 +220,59 @@ function constructTagList(data) {
...
+    localStorage.removeItem('toDelete');
+  }
+  else {
+  memoriesWaitingToBeRemoved.memories.splice(index, 1);
+  const memoriesStillToDelete = JSON.stringify(memoriesWaitingToBeRemoved);
+  localStorage.toDelete = memoriesStillToDelete;
+}
+}
+
+function removeMemoriesDeletedOffline(cb) {
```
Comment (des-des) [inline src/client/helpers/helpers.js]: `deletePendingMemories`

File: src/client/helpers/helpers.js [hunk 12]
```diff
@@ -220,6 +220,59 @@ function constructTagList(data) {
...
+  if(localStorage.getItem('toDelete') !== null) {
+  const deletedMemories = JSON.parse(localStorage.getItem('toDelete'));
+  deletedMemories.memories.forEach((memory, index) => {
```
Comment (des-des) [inline src/client/helpers/helpers.js]: This is bad..

You should use `async.parrallel`
https://caolan.github.io/async/docs.html#parallel

If you do not feel confident doing this dw

**Review by des-des — CHANGES_REQUESTED** [review]

File: src/client/helpers/helpers.js (hunk 12 above)
Comment (pbywater) [reply src/client/helpers/helpers.js]: I'm not really too sure how to implement this here!

**Review by des-des — APPROVED** [review]

**Pull request merged**